### PR TITLE
Send rejection/approval emails to authors of items sent in review - Fix

### PR DESCRIPTION
**Pull request by SudhirKumar2254**
Send rejection/approval emails to authors of items sent in review

### CODE REVIEW

File: items/Project/Global/System.Workflows/Demo Workflow/Awaiting Approval/Approve/Submission Notification.yml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      Value: "Demo Shared: Items Approved"
+    - ID: "5dd74568-4d4b-44c1-b513-0af5f4cda34f"
+      Hint: __Created by
+      Value: |
+        sitecore\admin
+    - ID: "e8f76062-0990-4b8b-860e-de4f50eb76ea"
+      Hint: Message
+      Value: |
+        This message is generated by a demo workflow action and was sent to the user who requested for the item approval. <br />
+        The following item was approved by $actionedby$: <br /><br />
```
Comment (jflheureux) [inline items/Project/Global/System.Workflows/Demo Workflow/Awaiting Approval/Approve/Submission Notification.yml:39]: Other tokens are using camelCase where words except the first one have a capitalized first letter. Could this token please be renamed to `actionedBy` everywhere it is used?

File: items/Project/Global/System.Workflows/Demo Workflow/Awaiting Approval/Approve/Submission Notification.yml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        Comments: $comment$ <br />
+        Item path : $itemPath$ <br />
+        Item language : $itemLanguage$ <br />
+        Item Version : $itemVersion$ <br /><br />
```
Comment (jflheureux) [inline items/Project/Global/System.Workflows/Demo Workflow/Awaiting Approval/Approve/Submission Notification.yml:43]: The `<br />`s at the end of this last line are not necessary. Could they please be removed?

File: items/Project/Global/System.Workflows/Demo Workflow/Awaiting Approval/Reject/Submission Notification.yml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        sitecore\admin
+    - ID: "e8f76062-0990-4b8b-860e-de4f50eb76ea"
+      Hint: Message
+      Value: |
+        This message is generated by a demo workflow action and was sent to the user who requested for the item approval. <br />
+        The following item was rejected by $actionedby$: <br /><br />
+        Comments: $comment$ <br />
+        Item path : $itemPath$ <br />
+        Item language : $itemLanguage$ <br />
+        Item Version : $itemVersion$ <br /><br />
```
Comment (jflheureux) [inline items/Project/Global/System.Workflows/Demo Workflow/Awaiting Approval/Reject/Submission Notification.yml:43]: The `<br />`s at the end of this last line are not necessary. Could they please be removed?

File: src/Foundation/Workflow/code/Actions/SubmissionNotification.cs [hunk 4]
```diff
@@ -10,124 +10,129 @@
...
-            {
-                Log.Error($"No Users with valid email addresses found in role {roleName}", typeof(SubmissionNotification));
-            }
-
-            return mailMessage;
-        }
-    }
-}
\ No newline at end of file
+	public class SubmissionNotification
```
Comment (jflheureux) [inline src/Foundation/Workflow/code/Actions/SubmissionNotification.cs:13]: All 4 spaces indentation were replaced by tab characters in this file. Could you please revert to 4 spaces indentation?

File: src/Foundation/Workflow/code/Actions/SubmissionNotification.cs [hunk 5]
```diff
@@ -10,124 +10,129 @@
...
+					.Where(x => x.IsInRole(role))
+					.Where(user => !string.IsNullOrEmpty(user.Profile.Email))
+					.ToList();
+
+				foreach (User user in users)
+				{
+					mailMessage.To.Add(user.Profile.Email);
+				}
+			}
+			else if (User.Exists(roleName))
```
Comment (jflheureux) [inline src/Foundation/Workflow/code/Actions/SubmissionNotification.cs:125]: It feels weird to check whether a role exists in users. I suggest we rename the `roleName` variable to `principalName` in this method.

File: src/Foundation/Workflow/code/Actions/SubmissionNotification.cs [hunk 6]
```diff
@@ -10,124 +10,129 @@
...
+			{
+				User user = User.FromName(roleName, false);
```
Comment (jflheureux) [inline src/Foundation/Workflow/code/Actions/SubmissionNotification.cs:127]: Can the `user` object be null? We should add a check for that.
Also, we should add a check whether the user has an email address configured in its profile before adding it to the recipients.

File: src/Foundation/Workflow/code/Actions/SubmissionNotification.cs [hunk 7]
```diff
@@ -10,124 +10,129 @@
...
+				mailMessage.To.Add(user.Profile.Email);
+			}
+			else
+			{
+				Log.Error($"No Users with valid email addresses found in role {roleName}", typeof(SubmissionNotification));
```
Comment (jflheureux) [inline src/Foundation/Workflow/code/Actions/SubmissionNotification.cs:132]: As we change `roleName` to `principalName`, we should update this message to `No users with valid email addresses found for security principal {principalName}`

**Review by jflheureux — CHANGES_REQUESTED** [review]
Very good work. The end result is as expected. There are however a couple of things to improve in the code and items. Thanks for your contribution

File: items/Project/Global/System.Workflows/Demo Workflow/Awaiting Approval/Approve/Submission Notification.yml [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+- Language: "ja-JP"
```
[the remark this answers is not in the document]
Comment (SudhirKumar2254) [reply items/Project/Global/System.Workflows/Demo Workflow/Awaiting Approval/Approve/Submission Notification.yml:44]: Thanks Jeff for your feedback. This Japanese version exists already in code. But I will delete the Japanese version from email actions in Demo workflows.

File: items/Project/Global/System.Workflows/Demo Workflow/Awaiting Approval/Reject/Submission Notification.yml [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+- Language: "ja-JP"
```
[the remark this answers is not in the document]
Comment (SudhirKumar2254) [reply items/Project/Global/System.Workflows/Demo Workflow/Awaiting Approval/Reject/Submission Notification.yml:44]: Sure will delete the Japanese version from email actions in workflows.

File: items/Project/Global/System.Workflows/Demo Workflow/Awaiting Approval/Approve/Submission Notification.yml (hunk 1 above)
Comment (SudhirKumar2254) [reply items/Project/Global/System.Workflows/Demo Workflow/Awaiting Approval/Approve/Submission Notification.yml:39]: Sure will do it

File: src/Foundation/Workflow/code/Actions/SubmissionNotification.cs (hunk 5 above)
Comment (SudhirKumar2254) [reply src/Foundation/Workflow/code/Actions/SubmissionNotification.cs:125]: Sure will rename the variable name.

File: src/Foundation/Workflow/code/Actions/SubmissionNotification.cs (hunk 6 above)
Comment (SudhirKumar2254) [reply src/Foundation/Workflow/code/Actions/SubmissionNotification.cs:127]: Null check is already there since User.Exists will check for null object.
However for email address for a profile, I will check for its existence.

File: src/Foundation/Workflow/code/Actions/SubmissionNotification.cs (hunk 7 above)
Comment (SudhirKumar2254) [reply src/Foundation/Workflow/code/Actions/SubmissionNotification.cs:132]: sure will do it.

Comment (SudhirKumar2254) [reply src/Foundation/Workflow/code/Actions/SubmissionNotification.cs:132]: Thanks for all your valuable feedback

File: items/Project/Global/System.Workflows/Demo Workflow/Awaiting Approval/Approve/Submission Notification.yml (hunk 2 above)
Comment (SudhirKumar2254) [reply items/Project/Global/System.Workflows/Demo Workflow/Awaiting Approval/Approve/Submission Notification.yml:43]: will do it

File: items/Project/Global/System.Workflows/Demo Workflow/Awaiting Approval/Reject/Submission Notification.yml (hunk 3 above)
Comment (SudhirKumar2254) [reply items/Project/Global/System.Workflows/Demo Workflow/Awaiting Approval/Reject/Submission Notification.yml:43]: will do it

File: items/Project/Global/System.Workflows/Demo Workflow/Draft/Submit/Submission Notification.yml [hunk 10]
```diff
@@ -38,30 +38,6 @@ Languages:
         Comments: $comment$ <br />
         Item path : $itemPath$ <br />
         Item language : $itemLanguage$ <br />
-        Item Version : $itemVersion$ <br /><br />
+        Item Version : $itemVersion$ <br />
```
Comment (jflheureux) [inline items/Project/Global/System.Workflows/Demo Workflow/Draft/Submit/Submission Notification.yml:41]: For this item, The dual `<br />` is important to add a blank line between the properties block and the next paragraph. Please revert this change.

File: src/Foundation/Workflow/code/Actions/SubmissionNotification.cs [hunk 11]
```diff
@@ -10,124 +10,132 @@
...
+			{
+				User user = User.FromName(principalName, false);
+				if (!string.IsNullOrEmpty(user.Profile.Email))
+				{
+					mailMessage.To.Add(user.Profile.Email);
+				}
+			}
+			else
+			{
+				Log.Error($"No Users with valid email addresses found in role {principalName}", typeof(SubmissionNotification));
```
Comment (jflheureux) [inline src/Foundation/Workflow/code/Actions/SubmissionNotification.cs:135]: Please change `role` to `security principal` in this message.

Comment (jflheureux) [reply src/Foundation/Workflow/code/Actions/SubmissionNotification.cs:135]: Please also change `Users` to `users` (lowercase)

File: src/Foundation/Workflow/code/Actions/SubmissionNotification.cs (hunk 4 above)
Comment (jflheureux) [reply src/Foundation/Workflow/code/Actions/SubmissionNotification.cs:13]: This is still an issue with the pull request. When you used the Visual Studio extension to remove trailing whitespace at the end of lines, the file was formatted with whatever formatting rules are configured in your copy of Visual Studio. This changed the indentation of all the lines of this file from "4 spaces" to "tab" characters. Please convert the lines indentation back to "4 spaces" instead of "tab" characters. Thanks

**Review by jflheureux — CHANGES_REQUESTED** [review]

File: items/Project/Global/System.Workflows/Demo Workflow/Draft/Submit/Submission Notification.yml (hunk 10 above)
Comment (SudhirKumar2254) [reply items/Project/Global/System.Workflows/Demo Workflow/Draft/Submit/Submission Notification.yml:41]: sure will revert the change

File: src/Foundation/Workflow/code/Actions/SubmissionNotification.cs (hunk 11 above)
Comment (SudhirKumar2254) [reply src/Foundation/Workflow/code/Actions/SubmissionNotification.cs:135]: will do

File: src/Foundation/Workflow/code/Actions/SubmissionNotification.cs (hunk 4 above)
Comment (SudhirKumar2254) [reply src/Foundation/Workflow/code/Actions/SubmissionNotification.cs]: will do .

**Review by jflheureux — APPROVED** [review]
